controller_manager:
  ros__parameters:
    update_rate: 100
    use_sim_time: true

    # RESPONSIBLE FOR PUBLISHING THE CURRENT STATE OF EACH JOINT INTO A TOPIC
    joint_broad:
      type: joint_state_broadcaster/JointStateBroadcaster
      
    velocity_controller:
      type: velocity_controllers/JointGroupVelocityController
      
    diff_cont:
      type: diff_drive_controller/DiffDriveController

diff_cont:
  ros__parameters:
    type: "diff_drive_controller/DiffDriveController"
    use_stamped_vel: false
    left_wheel_names: ["left_front_wheel_hinge","left_back_wheel_hinge"]
    right_wheel_names: ["right_front_wheel_hinge","right_back_wheel_hinge"]
    
    wheel_separation: 0.35
    wheels_per_side: 2
    wheel_radius: 0.05

    # Wheel separation and radius multipliers
    wheel_separation_multiplier: 1.0 # default: 1.0
    left_wheel_radius_multiplier: 1.0 # default: 1.0
    right_wheel_radius_multiplier: 1.0 # default: 1.0

    odom_frame_id: odom
    base_frame_id: base_link

    pose_covariance_diagonal: [0.001, 0.001, 1000000.00, 1000000.00, 1000000.00, 0.03]
    twist_covariance_diagonal: [0.001, 0.001, 0.001, 1000000.00, 1000000.00, 0.03]
    cmd_vel_timeout: 0.25

    # Odometry fused with IMU is published by robot_localization, so
    # no need to publish a TF based on encoders alone.
    enable_odom_tf: true

    publish_limited_velocity: true
    publish_wheel_data: true
    publish_rate: 50.0
    # velocity_rolling_window_size: 10
    # Velocity and acceleration limits
    # Whenever a min_* is unspecified, default to -max_*
    linear:
      x:
        has_velocity_limits: true
        max_velocity: 2.00   # m/s
        min_velocity: -0.5
        has_acceleration_limits: true
        max_acceleration: 0.80   # m/s^2
        min_acceleration: -0.40
        has_jerk_limits: false
    angular:
      z:
        has_velocity_limits: true
        max_velocity: 1.7   # rad/s
        min_velocity: -1.7
        has_acceleration_limits: true
        max_acceleration: 1.5   # rad/s^2
        min_acceleration: -1.5
        has_jerk_limits: false
      
velocity_controller:
  ros__parameters:
    joints:
     - left_back_wheel_hinge
     - right_back_wheel_hinge
     - left_front_wheel_hinge
     - right_front_wheel_hinge

joint_broad:
  ros__parameters:
    type: "joint_state_broadcaster/JointStateBroadcaster"
    joints:
     - left_back_wheel_hinge
     - right_back_wheel_hinge
     - left_front_wheel_hinge
     - right_front_wheel_hinge